### PR TITLE
Consider attribute_shop association when generating combinations

### LINKED ISSUE #30969: I generate combination in multishop with attributes not assiociated to all the shops
**Pb**

In multistore, when I apply the all store during the generation of the combination, some attributes do not belong to all the shops.

To avoid the creation of hazardous combination, the "apply to all shop" during the generation works only if the selected attributes are well associated to all the stores linked to this product

**Solution**

After checking "apply to all shops" and clicked "generate"  then an error message is displayed showing the list of attributes that are not associated with all stores when at least one of the attribute is not linked

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | https://github.com/PrestaShop/PrestaShop/issues/30969. When generating combinations for all shops, the generator will now assert if all the attribute groups and attributes exists in all of the shops before proceeding. If it at least one attribute is not associated to all shops, then it will show error. Also generation modal now only has attributes that exist in  current context shop (before it used to show all the attributes from generi attribute table). For now error is simple growl notification, in future we can improve the UX if needed. Screenshot 2023-01-20 at 17 30 25
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | TBD
| Fixed ticket?     | https://github.com/PrestaShop/PrestaShop/issues/30969
| Related PRs       | 
| Sponsor company   |

### CODE REVIEW

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php
Comment (zuk3975) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php]: A new wording here. Ping @PrestaShop/wording-managers for adjustments.
The case - when we try to generate combinations in all shops, but the attributes between shops are different, then this error is thrown.

File: src/Adapter/Product/AttributeGroup/QueryHandler/GetProductAttributeGroupsHandler.php
Comment (jolelievre) [inline src/Adapter/Product/AttributeGroup/QueryHandler/GetProductAttributeGroupsHandler.php]: To be consistent, shouldn't the `$attributeIds = $this->attributeRepository->getProductAttributesIds($query->getProductId());` also be based on the shop constraint?

File: src/Core/Domain/Product/Combination/Exception/CannotGenerateCombinationException.php
Comment (jolelievre) [inline src/Core/Domain/Product/Combination/Exception/CannotGenerateCombinationException.php]:
```suggestion
    public const DIFFERENT_ATTRIBUTES_BETWEEN_SHOPS = 1;
```

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php
Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php]:
```suggestion
                    'Selected attributes must be associated with every store to generate a combination for all stores',
```

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php]: Alternative
```suggestion
                    'You cannot generate combinations for all stores if the selected attributes are not associated on all stores',
```

File: src/PrestaShopBundle/Entity/Repository/AttributeGroupRepository.php
Comment (jolelievre) [inline src/PrestaShopBundle/Entity/Repository/AttributeGroupRepository.php]: The new ShopConstraint is missing here

File: src/PrestaShopBundle/Entity/Repository/AttributeGroupRepository.php
Comment (jolelievre) [inline src/PrestaShopBundle/Entity/Repository/AttributeGroupRepository.php]:
```suggestion
                throw new ShopAssociationNotFound('Provided attribute groups do not exist in every shop');
```

File: src/PrestaShopBundle/Entity/Repository/AttributeGroupRepository.php
Comment (jolelievre) [inline src/PrestaShopBundle/Entity/Repository/AttributeGroupRepository.php]:
```suggestion
            throw new ShopAssociationNotFound('Provided attribute groups do not exist in every shop');
```

File: src/PrestaShopBundle/Entity/Repository/AttributeGroupRepository.php
Comment (jolelievre) [inline src/PrestaShopBundle/Entity/Repository/AttributeGroupRepository.php]: I'm not sure to understand this query nor the checks you made 🤔
I would have thought of a more simple implementation:
- query with join to get all the attribute groups IDs and their associated shop_id from `attribute_group_shop`
- loop through results and store in an associative array `$attributeGroupShops[$result['id_attribute_group']][] = $result['id_shop']` thus you get the exact list of shops for each attribute group
- finally loop through this associate array and check that for each attribute groups the list of shop IDs is equal to `$shopIdValues`

This way the implementation would be easier to understand, and it not only checks the could but also the associated values.

File: src/PrestaShopBundle/Entity/Repository/AttributeRepository.php
Comment (jolelievre) [inline src/PrestaShopBundle/Entity/Repository/AttributeRepository.php]: Same suggestion as for attribute group

File: src/PrestaShopBundle/Resources/config/services/bundle/repository.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/repository.yml]: Can you please put this new service declaration after `prestashop.core.admin.attribute_group.repository` The idea is to keep the alias close to its original service

File: tests/Integration/Behaviour/Features/Context/Domain/AttributeGroupFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/AttributeGroupFeatureContext.php:47]: I know it represents quite some work, but since you added the ShopConstraint to this query maybe it would be worth adding some tests for specific shop and for all shops You can't really handle the creation with current steps so it would always be associated to the default shop

But since you have a step to handle the association between Attribute group and its shops, you create two groups, associate the first to shop2 the second to shop3 And check that depending on the shop the list is different, but when you get the list for all shops then they are all present

**Review by jolelievre — COMMENTED** [review]
Thanks @zuk3975 

A few comments mostly details, except:
- the implementation of the assertExistInAllShops methods seems confusing to me (but maybe it's just me) so I suggested an alternative implementation, don't hesitate to show it to other developers maybe they get it right then no need for modifications
- since you added the shop constraint in some of the queries for attribute groups, they should be tested in behat as well

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php
Comment (zuk3975) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php]: still waiting for @PrestaShop/wording-managers decision approval before changes (btw the second suggestion doesn't sound well with the "on all stores") 😄

File: src/PrestaShopBundle/Entity/Repository/AttributeGroupRepository.php
Comment (jolelievre) [inline src/PrestaShopBundle/Entity/Repository/AttributeGroupRepository.php]: Maybe this should be moved in the other repository, especially since you use `getArrayResult` so if it's not relying on the Entity it might as well be in the ObjectModel/Connection repository instead?

File: src/PrestaShopBundle/Entity/Repository/AttributeRepository.php
Comment (jolelievre) [inline src/PrestaShopBundle/Entity/Repository/AttributeRepository.php]: Same suggestion it seems like it can easily be moved in the other repository Let's avoid this Doctrine EntityRepository as much as we can 😅

File: src/PrestaShopBundle/Resources/config/services/bundle/repository.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/repository.yml]: If you move the method you added in this repo in the other one I believe you can clean this definition

**Review by jolelievre — DISMISSED** [review]
Thanks @zuk3975

**Review by lartist — DISMISSED** [review]

**Review by djoelleuch — COMMENTED** [review]
Hello @zuk3975 ,

I have checked your PR , and the issue is still exist .

[Untitled_ Jan 30, 2023 4_18 PM.webm](https://user-images.githubusercontent.com/115152460/215533203-7c622dc4-9ed4-456c-9618-809db788022d.webm)

could you please check it .
Thank you !

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php
Comment (l-delin) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php]:
```suggestion
                    'To create combinations for all your stores, the selected attributes must be available on each of them.'
```

Comment (l-delin) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php]: Hello @zuk3975, sorry for the late review, I was away.

**Comment by jolelievre** [comment]
Hi @zuk3975 this one also has some new conflicts after my PR as merged

**Review by jolelievre — DISMISSED** [review]

**Comment by zuk3975** [comment]
@djoelleuch I tried to test it again and it seems to work just fine 🤔 (maybe by accident you checked on wrong branch or something? 🤷‍♂️ ). Adding my video record bellow
https://user-images.githubusercontent.com/31609858/216061947-6055ac5c-b66f-4565-8fb3-0545ac22f2a7.mov

**Review by jolelievre — DISMISSED** [review]
Thanks @zuk3975

**Review by lartist — DISMISSED** [review]

**Review by jolelievre — APPROVED** [review]

**Review by lartist — APPROVED** [review]

**Review by HanaRebaiQA — APPROVED** [review]
Hello @zuk3975 

I have checked your PR. The feature is OK.

-Check the automated tests: OK
-Check with different combinations of attributes in each shop: OK
-Check with a new shop and a new group of shops: OK
-Check when assigning the attributes for all shops  OK
-Check with different browsers: OK


image

https://user-images.githubusercontent.com/92912932/216647375-73080b40-2550-43b3-bb28-d6229b51a7b8.mp4

So, It's QA :heavy_check_mark: 

Thank You!

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php
Comment (l-delin) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php:758]:
```suggestion
                    'Make sure the selected attributes are associated with all your stores to create this combination.',
```

**Comment by jolelievre** [comment]
Thanks for the feedback @l-delin 
Since the PR is validated by QA we merge it to validate the functionality, I leave the label `Waiting for wording` so that we don't forget to fix it in another PR though

**Comment by jolelievre** [comment]
Thanks @zuk3975

**Pull request merged**
